pydantic/dataclasses.py

```diff
@@ -103,76 +105,74 @@ def __validate__(cls: Type['DataclassT'], v: Any) -> 'DataclassT':
-        frozen: bool = False,
-        config: Union[ConfigDict, Type[object], None] = None,
-        validate_on_init: Optional[bool] = None,
-    ) -> Callable[[Type[_T]], 'DataclassClassOrWrapper']:
-        ...
-
-    @dataclass_transform(kw_only_default=True, field_specifiers=(Field, FieldInfo))
-    @overload
-    def dataclass(
-        _cls: Type[_T],
-        *,
-        init: bool = True,
-        repr: bool = True,
-        eq: bool = True,
-        order: bool = False,
-        unsafe_hash: bool = False,
-        frozen: bool = False,
-        config: Union[ConfigDict, Type[object], None] = None,
-        validate_on_init: Optional[bool] = None,
-    ) -> 'DataclassClassOrWrapper':
-        ...
-
-
-@dataclass_transform(kw_only_default=True, field_specifiers=(Field, FieldInfo))
+if TYPE_CHECKING:
+    if sys.version_info >= (3, 10):
+
+        @dataclass_transform(field_specifiers=(dataclasses.field, Field))
+        @overload
+        def dataclass(
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Union[ConfigDict, Type[object], None] = None,
+            validate_on_init: Optional[bool] = None,
+            kw_only: bool = ...,
+        ) -> Callable[[Type[_T]], 'DataclassClassOrWrapper']:
+            ...
+
+        @overload
+        def dataclass(
+            _cls: Type[_T],
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Union[ConfigDict, Type[object], None] = None,
+            validate_on_init: Optional[bool] = None,
+            kw_only: bool = ...,
+        ) -> 'DataclassClassOrWrapper':
+            ...
+
+    else:
+
+        @dataclass_transform(field_specifiers=(dataclasses.field, Field))
+        @overload
+        def dataclass(
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Union[ConfigDict, Type[object], None] = None,
+            validate_on_init: Optional[bool] = None,
+        ) -> Callable[[Type[_T]], 'DataclassClassOrWrapper']:
+            ...
+
+        @overload
+        def dataclass(
+            _cls: Type[_T],
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Union[ConfigDict, Type[object], None] = None,
+            validate_on_init: Optional[bool] = None,
+        ) -> 'DataclassClassOrWrapper':
+            ...
+
+
 def dataclass(
```

Comment:
Should the implementation keep the `dataclass_transform` for runtime introspection?

Reply:
In that case we would need to import stdlib `dataclasses` at runtime. The goal was to avoid just that.
Unfortunately, it doesn't seem to be valid to use forward references as value for `field_specifiers`. Although mypy and pyright seems to know how to handle it already 🤔

Not sure how to processed here if I'm honest. I could take another look at the import situation but probably not until later this week unfortunately.